### PR TITLE
Expose an option for producing 64-bit DWARF format

### PR DESCRIPTION
**Pull request by davidgm94**
This commit enables producing 64-bit DWARF format for Zig executables that are produced through the LLVM backend. This is achieved by exposing both command-line flags and CompileStep flags. The production of the 64-bit format only affects binaries that use the DWARF format and it is disabled on MacOS due to it being problematic. This commit, despite generating the interface for the Zig user to be able to tell the compile which format is wanted, is just implemented for the LLVM backend, so clang and the self-hosted backends will need this to be implemented in a future commit.

This is an effort to work around #7962, since the emission of the 64-bit format automatically produces 64-bit relocations. Further investigation will be needed to make DWARF 32-bit format to emit bigger relocations when needed and not make the linker angry.

### CODE REVIEW

**Comment by davidgm94** [comment]
I must say that I tested poorly the 64-bit relocations and this does not solve the problem. I tested by placing debug info sections in a separate program segment, which made the base address 0x0, so this problem never could have happened. By placing the sections again in one with the typical kernel higher half offset (0xffff_ffff_8000_0000) the error still happens. Let me know if you still want to support 64-bit DWARF.

**Comment by kubkon** [comment]
Also, last time I checked DWARF64 is not well-supported by lldb in general, so we might have problems on Linux too.

**Comment by davidgm94** [comment]
Yeah, totally, that would save the whole interface stuff, but... is it safe to do? Won't there be bugs or debugger incompatibilities?

**Comment by kubkon** [comment]
> Yeah, totally, that would save the whole interface stuff, but... is it safe to do? Won't there be bugs or debugger incompatibilities?

That's a good point. A flag might be a good idea for now.

File: src/codegen/llvm.zig [hunk 1]
```diff
@@ -428,7 +428,18 @@ pub const Object = struct {
         if (!options.strip) {
             switch (options.target.ofmt) {
                 .coff => llvm_module.addModuleCodeViewFlag(),
-                else => llvm_module.addModuleDebugInfoFlag(),
+                else => {
+                    const dwarf_format: std.builtin.DwarfFormat = switch (options.target.isDarwin()) {
+                        // For Darwin targets, using DWARF64 is problematic, so stick to DWARF32
+                        true => .dwarf32,
```
Comment (kubkon) [inline src/codegen/llvm.zig:434]: To be honest, we don't have to special case Darwin as by default we use dwarf32, and I would assume the user would know the requirements of each platform if they override it using the flag so let's let them.

**Review by kubkon — CHANGES_REQUESTED** [review]

File: src/codegen/llvm.zig (hunk 1 above)
Comment (davidgm94) [reply src/codegen/llvm.zig]: Done :)

**Review by kubkon — APPROVED** [review]
Looks good to me! @andrewrk could you have a look also please?

File: lib/std/builtin.zig [hunk 2]
```diff
@@ -462,6 +462,13 @@ pub const WasiExecModel = enum {
     reactor,
 };
 
+/// This data structure is used by the Zig language code generation and
+/// therefore must be kept in sync with the compiler implementation.
+pub const DwarfFormat = enum {
+    dwarf32,
+    dwarf64,
+};
+
```
Comment (andrewrk) [inline lib/std/builtin.zig:471]: Why is this in std.builtin? Seems like it belongs in `std.dwarf`, no?

Comment (davidgm94) [reply lib/std/builtin.zig]: Oh I didn't know where to put it, but yeah, it makes sense. Fixed.

**Pull request merged**